src/hazelcore/OcamlExtraction.re

```diff
@@ -0,0 +1,374 @@
+  let rec extract = (~dp: DHPat.t): ocaml_pat => {
+    switch (dp) {
+    | EmptyHole(_, _) => failwith("Pat: Empty Hole")
+    | NonEmptyHole(_, _, _, _) => failwith("Pat: NonEmptyHole")
+    | Wild => "_"
+    | Keyword(_, _, _) => failwith("Pat: Incomplete Program, Keyword")
+    | Var(s) => s
+    | IntLit(i) => string_of_int(i)
+    | FloatLit(f) => string_of_float(f)
+    | BoolLit(b) => string_of_bool(b)
+    | Inj(_side, t) => extract(~dp=t)
+    | ListNil => "[]"
+    | Cons(dp1, dp2) => extract(~dp=dp1) ++ "::" ++ extract(~dp=dp2)
+    | Pair(dp1, dp2) =>
+      "(" ++ extract(~dp=dp1) ++ ", " ++ extract(~dp=dp2) ++ ")"
+    | Triv => "()"
+    | Ap(dp1, dp2) => extract(~dp=dp1) ++ " " ++ extract(~dp=dp2)
+    };
+  };
+};
+
+module Exp = {
+  type ocaml_exp = string;
+
+  // (extracted expression, dynamic/inner type) like ExpandResult
+  // Htyp.t is like the ground type
+
+  //type extract_result = option((ocaml_exp, HTyp.t));
+  type extract_result = (ocaml_exp, HTyp.t);
+
+  /* Theorem: If extract(ctx, d) = Some(ce, ty) and ctx |- d : ty then ce :_ocaml Typ.extract(ty) */
+
+  //TODO: replace == with HTyp.consistent
+  let rec extract = (~ctx: Contexts.t, ~de: DHExp.t): extract_result =>
+    switch (de) {
+    | EmptyHole(_) => failwith("Exp: Empty Hole")
+    | NonEmptyHole(_) => failwith("Exp: Non-Empty Hole")
+    // used for variables that are keywords in Hazel, like "let" and "case"
+    | Keyword(_) => failwith("Exp: Incomplete Program (Keyword)")
+    | FreeVar(_, _, map, x) =>
+      //lookup the expression represented by x
+      let map_e = VarMap.lookup(map, x);
+      switch (map_e) {
+      | None => failwith("Exp: FreeVar Not Found")
+      | Some(e) =>
+        let ocaml_e = extract(~ctx, ~de=e);
+        //add a parenthesize to avoid any possible order priority
+        (
+          "(" ++ x ++ " : " ++ Typ.extract(~t=snd(ocaml_e)) ++ ")",
+          snd(ocaml_e),
+        );
+      };
+    | BoundVar(x) =>
+      //directly lookup type in environment
+      let typ = VarCtx.lookup(Contexts.gamma(ctx), x);
+      switch (typ) {
+      | None => failwith("Exp: BoundVar " ++ x ++ " Not Found")
+      // if the hole type, we don't print the type
+      | Some(t) =>
+        if (t == Hole) {
+          ("(" ++ x ++ ")", t);
+        } else {
+          ("(" ++ x ++ " : " ++ Typ.extract(~t) ++ ")", t);
+        }
+      };
+    | Let(dp, de1, de2) =>
+      switch (dp) {
+      | Var(x) =>
+        let ocaml_dp = Pat.extract(~dp);
+        let ocaml_de1 = extract(~ctx, ~de=de1);
+        let ctx' = Contexts.extend_gamma(ctx, (x, snd(ocaml_de1)));
+        let ocaml_de2 = extract(~ctx=ctx', ~de=de2);
+        // use the "let .. = ..;;" format
+        // if it's a recursive function, we skip to directly evaluating the FixF (de1)
+        // because the recursive function should be declared by "rec" only in ocaml
+        switch (de1) {
+        | FixF(_, _, _) => (
+            fst(ocaml_de1) ++ fst(ocaml_de2),
```

Comment:
this should turn into a let rec here I think, not in the FixF case below

Reply:
May I ask how the FixF will appear? Will it appear as Let(..., FixF(...), ...) or directly FixF if a recursive function is declared? Here when the Let encounters an FixF, I directly pass it to FixF handler to extract to "let rec".

Reply:
Hazel programs will only ever result in internal expressions containing let ... = fix ..., so you can assume that.